src/slot2/GBACart.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+    Copyright (c) 2019 Adrian "asie" Siekierka
+
+    This file is part of melonDS.
+
+    melonDS is free software: you can redistribute it and/or modify it under
+    the terms of the GNU General Public License as published by the Free
+    Software Foundation, either version 3 of the License, or (at your option)
+    any later version.
+
+    melonDS is distributed in the hope that it will be useful, but WITHOUT ANY
+    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with melonDS. If not, see http://www.gnu.org/licenses/.
+*/
+
+#include <stdio.h>
+#include <string.h>
+#include "../Platform.h"
+#include "GBACart.h"
+#include "GBACartNone.h"
+
+#define RAM_TO_CART(addr) (((addr) >> 1) & 0x00FFFFFF)
+
+namespace GBACartHelper {
+    static GBACart *cart = &GBACartNone::singleton();
+
+    void Use(GBACart *inputCart) {
```

Comment:
follow up to my suggestion of moving the cart selection into the core. Then we don't need to move raw pointer through so many places of the code (frontend -> NDS.cpp -> here) and instead they stay here contained. Then we also don't need this singleton thing, since we can control the instantiation from a single place.

Reply:
The main reason it's done this way is because each cart can take different arguments (for example: a CF/SD image path, a cartridge ROM path, et cetera). Do you have a better proposal on how to handle this? I have some ideas (I'd pass a struct containing all possible configuration options), but would like to hear yours if they differ.

Reply:
I see, the gui needs to communicate with the individual cart types. Maybe we could have a function per carttype named UseCartXYOrSo where all of the object construction and error checking is done, so only the switch, which passes the data from the control elements to the appropriate function remains in the frontend.

That all said, we still need to save the config somehow. Maybe it would be the smartest to add all the settings as config entries. Then only function which updates the cartridge configuration would be needed. It would take the config data by itself to construct the appriate object with the right settings. The gui dialog would then only have to change the right config data and call the "update" method.

Reply:
I've gone for the struct + config entries approach, if you could have a look and let me know what you think I'd appreciate that.

Reply:
I was a bit confused, but after going through your commit, I saw you've already implemented it. 👍 It solves all problems I mentioned, looks good to me